// react-explode/Explosion2.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _gsap = require("gsap");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Explosion = function (_Component) {
    _inherits(Explosion, _Component);

    function Explosion() {
        var _ref;

        var _temp, _this, _ret;

        _classCallCheck(this, Explosion);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Explosion.__proto__ || Object.getPrototypeOf(Explosion)).call.apply(_ref, [this].concat(args))), _this), _this.state = { size: 400, delay: 0, repeatDelay: 0, repeat: 0 }, _this.timeline = null, _this.radius = 47.5, _this.strokeWidth = 1, _this.explode = function () {
            var _this$state = _this.state,
                size = _this$state.size,
                repeat = _this$state.repeat,
                repeatDelay = _this$state.repeatDelay,
                delay = _this$state.delay;
            var _this$props = _this.props,
                onStart = _this$props.onStart,
                onRepeat = _this$props.onRepeat,
                onComplete = _this$props.onComplete;

            var ease = _gsap.Power4.easeOut;
            var radius = size * _this.radius / 100;
            var strokeWidth = Math.ceil(size * _this.strokeWidth / 100);

            _this.timeline = new _gsap.TimelineMax({
                repeat: repeat,
                repeatDelay: repeatDelay,
                delay: delay,
                onStart: onStart,
                onComplete: onComplete,
                onRepeat: onRepeat
            });

            _this.timeline.fromTo(_this.circle, 1, { attr: { r: 0 } }, { attr: { r: radius }, ease: ease }).fromTo(_this.circle, 0.8, { attr: { "stroke-width": strokeWidth }, ease: ease }, { attr: { "stroke-width": 0 }, ease: ease }, "-=0.5").set(_this.circle, { attr: { "stroke-width": strokeWidth } }, "-=1");
        }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(Explosion, [{
        key: "componentDidMount",
        value: function componentDidMount() {
            this.explode();
        }
    }, {
        key: "componentDidUpdate",
        value: function componentDidUpdate() {
            this.timeline.kill();
            this.explode();
        }
    }, {
        key: "render",
        value: function render() {
            var _this2 = this;

            var size = this.state.size;
            var center = size / 2;
            var _props = this.props,
                style = _props.style,
                _props$color = _props.color,
                color = _props$color === undefined ? "white" : _props$color;

            var strokeWidth = Math.ceil(size * this.strokeWidth / 100);

            return _react2.default.createElement(
                "svg",
                { width: size, height: size, style: style },
                _react2.default.createElement("circle", {
                    cx: center,
                    cy: center,
                    r: 0,
                    strokeWidth: strokeWidth,
                    stroke: color,
                    fill: "none",
                    ref: function ref(el) {
                        return _this2.circle = el;
                    }
                })
            );
        }
    }], [{
        key: "getDerivedStateFromProps",
        value: function getDerivedStateFromProps(_ref2, prevState) {
            var size = _ref2.size,
                delay = _ref2.delay,
                repeatDelay = _ref2.repeatDelay,
                repeat = _ref2.repeat;

            if (size !== prevState.size || delay !== prevState.delay || repeatDelay !== prevState.repeatDelay || repeat !== prevState.repeat) return { size: size, delay: delay, repeatDelay: repeatDelay, repeat: repeat };

            return null;
        }
    }]);

    return Explosion;
}(_react.Component);

exports.default = Explosion;
